### PR TITLE
chain: Implement handler for data reports

### LINKED ISSUE #498: Implement handler for data requests and reports

### PR DESCRIPTION
**Pull request by taobun**
fixed #498

### CODE REVIEW

File: chain/x/zoracle/handler.go [hunk 1]
```diff
@@ -75,49 +75,21 @@ func handleMsgRequest(ctx sdk.Context, keeper Keeper, msg MsgRequestData) sdk.Re
...
-// 		sdk.NewEvent(
-// 			types.EventTypeReport,
-// 			sdk.NewAttribute(types.AttributeKeyRequestID, fmt.Sprintf("%d", msg.RequestID)),
-// 			sdk.NewAttribute(types.AttributeKeyCodeName, storedCode.Name),
-// 			sdk.NewAttribute(types.AttributeKeyValidator, msg.Validator.String()),
-// 		),
-// 	})
-// 	return sdk.Result{Events: ctx.EventManager().Events()}
-// }
+func handleMsgReport(ctx sdk.Context, keeper Keeper, msg MsgReportData) sdk.Result {
```
Comment (sorawit) [inline chain/x/zoracle/handler.go]: Let's call this function `handleMsgReportData` for the sake of consistency. Also update `handleMsgRequestData`

File: chain/x/zoracle/internal/keeper/report.go [hunk 2]
```diff
@@ -6,6 +6,68 @@ import (
...
+	}
+
+	if isFirstReport {
+		if int64(len(dataSet)) != k.GetRawDataRequestCount(ctx, requestID) {
+			// TODO: fix error later
+			return types.ErrRequestNotFound(types.DefaultCodespace)
+		}
+	}
+
+	lastExternalID := dataSet[0].ExternalDataID - 1
```
Comment (sorawit) [inline chain/x/zoracle/internal/keeper/report.go]: Careful of overflow!

File: chain/x/zoracle/internal/keeper/report.go [hunk 3]
```diff
@@ -6,6 +6,68 @@ import (
...
+		if validator.Equals(validValidator) {
+			found = true
+			break
+		}
+	}
+	if !found {
+		return types.ErrInvalidValidator(types.DefaultCodespace)
+	}
+
+	isFirstReport := true
```
Comment (sorawit) [inline chain/x/zoracle/internal/keeper/report.go]: This is complicated. Let's simplify by requiring that all raw data reports must come in one message.

File: chain/x/zoracle/internal/keeper/report.go [hunk 4]
```diff
@@ -6,6 +6,68 @@ import (
...
+	if isFirstReport {
+		if int64(len(dataSet)) != k.GetRawDataRequestCount(ctx, requestID) {
+			// TODO: fix error later
+			return types.ErrRequestNotFound(types.DefaultCodespace)
+		}
+	}
+
+	lastExternalID := dataSet[0].ExternalDataID - 1
+	for _, rawReport := range dataSet {
+		if lastExternalID >= rawReport.ExternalDataID {
```
Comment (sorawit) [inline chain/x/zoracle/internal/keeper/report.go]: Can we check that all raw requests must be reported for the transaction to be successful?

**Review by sorawit — CHANGES_REQUESTED** [review]

File: chain/x/zoracle/internal/keeper/report.go (hunk 4 above)
Comment (taobun) [reply chain/x/zoracle/internal/keeper/report.go]: The number of reports is checked  and if no report used same  externalID after check exist RawDataRequest, it guarantee that all raw requests must be reported.

**Review by sorawit — APPROVED** [review]

**Review by evilpeach — APPROVED** [review]

**Pull request merged**